prost-build/src/path.rs

```diff
@@ -1,61 +1,107 @@
 //! Utilities for working with Protobuf paths.
 
-use std::collections::HashMap;
 use std::iter;
 
 /// Maps a fully-qualified Protobuf path to a value using path matchers.
 #[derive(Debug, Default)]
 pub(crate) struct PathMap<T> {
-    matchers: HashMap<String, T>,
+    // insertion order might actually matter (to avoid warning about legacy-derive-helpers)
+    // see: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#legacy-derive-helpers
+    pub(crate) matchers: Vec<(String, T)>,
 }
 
 impl<T> PathMap<T> {
     /// Inserts a new matcher and associated value to the path map.
     pub(crate) fn insert(&mut self, matcher: String, value: T) {
-        self.matchers.insert(matcher, value);
+        self.matchers.push((matcher, value));
     }
 
-    /// Returns the value which matches the provided fully-qualified Protobuf path.
-    pub(crate) fn get(&self, fq_path: &'_ str) -> Option<&T> {
-        // First, try matching the full path.
-        iter::once(fq_path)
-            // Then, try matching path suffixes.
-            .chain(suffixes(fq_path))
-            // Then, try matching path prefixes.
-            .chain(prefixes(fq_path))
-            // Then, match the global path. This matcher must never fail, since the constructor
-            // initializes it.
-            .chain(iter::once("."))
-            .flat_map(|path| self.matchers.get(path))
-            .next()
+    pub(crate) fn get<'a>(&'a self, fq_path: &'_ str) -> Iter<'a, T> {
+        Iter::new(self, fq_path.to_string())
     }
 
-    /// Returns the value which matches the provided fully-qualified Protobuf path and field name.
-    pub(crate) fn get_field(&self, fq_path: &'_ str, field: &'_ str) -> Option<&T> {
-        let full_path = format!("{}.{}", fq_path, field);
-        let full_path = full_path.as_str();
-
-        // First, try matching the path.
-        let value = iter::once(full_path)
-            // Then, try matching path suffixes.
-            .chain(suffixes(full_path))
-            // Then, try matching path suffixes without the field name.
-            .chain(suffixes(fq_path))
-            // Then, try matching path prefixes.
-            .chain(prefixes(full_path))
-            // Then, match the global path. This matcher must never fail, since the constructor
-            // initializes it.
-            .chain(iter::once("."))
-            .flat_map(|path| self.matchers.get(path))
-            .next();
-
-        value
+    pub(crate) fn get_field<'a>(&'a self, fq_path: &'_ str, field: &'_ str) -> Iter<'a, T> {
+        Iter::new(self, format!("{}.{}", fq_path, field))
+    }
+
+    #[allow(unused)]
+    pub(crate) fn get_best<'a>(&'a self, fq_path: &'_ str) -> Option<&'a T> {
+        self.find_best_matching(fq_path)
+    }
+
+    pub(crate) fn get_best_field<'a>(&'a self, fq_path: &'_ str, field: &'_ str) -> Option<&'a T> {
```

Comment:
would a better name here be `get_first_field()` and might be worth it to document this behavior.

Reply:
Done